crates/ra_hir/src/ty.rs

```diff
@@ -907,14 +1057,34 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 self.infer_expr(*body, &Expectation::has_type(Ty::unit()));
                 Ty::unit()
             }
-            Expr::For { iterable, body, .. } => {
+            Expr::For {
+                iterable,
+                body,
+                pat,
+            } => {
                 let _iterable_ty = self.infer_expr(*iterable, &Expectation::none());
-                // TODO write type for pat
+                self.infer_pat(*pat, &Expectation::none());
                 self.infer_expr(*body, &Expectation::has_type(Ty::unit()));
                 Ty::unit()
             }
-            Expr::Lambda { body, .. } => {
-                // TODO write types for args, infer lambda type etc.
+            Expr::Lambda {
+                body,
+                args,
+                arg_types,
+            } => {
+                assert_eq!(args.len(), arg_types.len());
+
+                for (arg_pat, arg_type) in args.iter().zip(arg_types.iter()) {
+                    let expected = if let Some(tyref) = arg_type {
```

Comment:
totally a nit, but...
```suggestion
                    let expected = if let Some(type_ref) = arg_type {
```
since `TypeRef`s are always called that here ;)